plugins/account_history_plugin/account_history_plugin.cpp

```diff
@@ -0,0 +1,136 @@
+#include <eos/account_history_plugin/account_history_plugin.hpp>
+#include <eos/account_history_plugin/account_history_object.hpp>
+#include <eos/chain/chain_controller.hpp>
+#include <eos/chain/config.hpp>
+#include <eos/chain/exceptions.hpp>
+#include <eos/chain/transaction.hpp>
+#include <eos/chain/types.hpp>
+
+#include <fc/crypto/sha256.hpp>
+#include <fc/io/json.hpp>
+#include <fc/variant.hpp>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/mem_fun.hpp>
+#include <eos/chain/multi_index_includes.hpp>
+
+namespace fc { class variant; }
+
+namespace eos {
+
+using chain::block_id_type;
+using chain::ProcessedTransaction;
+using chain::signed_block;
+using boost::multi_index_container;
+using chain::transaction_id_type;
+using namespace boost::multi_index;
+
+class account_history_plugin_impl {
+public:
+   ProcessedTransaction get_transaction(const chain::transaction_id_type&  transaction_id) const;
+   void applied_block(const signed_block&);
+   chain_plugin* chain_plug;
+private:
+
+   optional<block_id_type> find_block_id(const transaction_id_type& transaction_id) const;
+};
+
+optional<block_id_type> account_history_plugin_impl::find_block_id(const transaction_id_type& transaction_id) const
+{
+   const auto& db = chain_plug->chain().get_database();
+   optional<block_id_type> block_id;
+   db.with_read_lock( [&]() {
+      const auto& trx_idx = db.get_index<transaction_history_multi_index, by_trx_id>();
+      auto transaction_history = trx_idx.find( transaction_id );
+      if (transaction_history != trx_idx.end())
+         block_id = transaction_history->block_id;
+   } );
+   return block_id;
+}
+
+ProcessedTransaction account_history_plugin_impl::get_transaction(const chain::transaction_id_type&  transaction_id) const
+{
+   auto block_id = find_block_id(transaction_id);
+   if( block_id.valid() )
+   {
+      auto block = chain_plug->chain().fetch_block_by_id(*block_id);
+      if (block.valid())
+      {
+         for (const auto& cycle : block->cycles)
+            for (const auto& thread : cycle)
+               for (const auto& trx : thread.user_input)
+                  if (trx.id() == transaction_id)
+                     return trx;
+      }
+
+      // ERROR in indexing logic
+      std::string msg = "transaction_id=" + transaction_id.str() + " indexed with block_id=" + block_id->str() + ", but ";
+      if (!block)
+         msg += "block was not found";
+      else
+         msg += "transaction was not found in the block";
+      BOOST_THROW_EXCEPTION( std::runtime_error( msg ) );
```

Comment:
Don't use a `std::runtime_error` -- we want to exclusively use our own exception types in this codebase.

This is throwing for a pretty weird case that, theoretically, should never actually occur, so we don't need to define an exception type for it specifically. We can just use assertions. Something like this should suffice:
```
FC_ASSERT(block, "Transaction with ID ${tid} was indexed as being in block ID ${bid}, but no such block was found", ("tid", transaction_id")("bid", block_id));
FC_THROW("Transaction with ID ${tid} was indexed as being in block ID ${bid}, but was not found in that block", ("tid", transaction_id")("bid", block_id));
```